### PR TITLE
Add pydantic support

### PR DESCRIPTION
**Pull request by steinitzu**
https://github.com/dlt-hub/dlt/issues/501

### CODE REVIEW

File: dlt/extract/decorators.py [hunk 1]
```diff
@@ -241,7 +242,7 @@ def resource(
     name: str = None,
     table_name: TTableHintTemplate[str] = None,
     write_disposition: TTableHintTemplate[TWriteDisposition] = None,
-    columns: TTableHintTemplate[TTableSchemaColumns] = None,
+    columns: TTableHintTemplate[TAnySchemaColumns] = None,
```
Comment (rudolfix) [inline dlt/extract/decorators.py:245]: please type dlt.transformer below

File: tests/extract/test_extract.py [hunk 2]
```diff
@@ -1,3 +1,5 @@
+from pydantic import BaseModel
```
Comment (rudolfix) [inline tests/extract/test_extract.py:1]: do we need this?

File: dlt/common/libs/pydantic.py [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+        if is_list_generic_type(inner_type):
+            inner_type = list
+        elif is_dict_generic_type(inner_type) or issubclass(inner_type, BaseModel):
+            inner_type = dict
+
+        name = field.alias or field_name
+
+        result[name] = {
+            "name": name,
+            "data_type": py_type_to_sc_type(inner_type),
```
Comment (rudolfix) [inline dlt/common/libs/pydantic.py:44]: 👍

File: dlt/common/libs/pydantic.py [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+
+    Returns:
+        TTableSchemaColumns: table schema columns dict
+    """
+    result: TTableSchemaColumns = {}
+
+    fields = model.__fields__
+    for field_name, field in fields.items():
+        annotation = field.annotation
+        nullable = is_optional_type(annotation)
```
Comment (rudolfix) [inline dlt/common/libs/pydantic.py:28]: 👍

File: dlt/common/libs/pydantic.py [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+from dlt.common.data_types import py_type_to_sc_type, TDataType
+from dlt.common.typing import is_optional_type, extract_inner_type, is_list_generic_type, is_dict_generic_type, is_union
+
+try:
+    from pydantic import BaseModel, Field
+except ImportError:
+    raise MissingDependencyException("DLT pydantic Helpers", ["pydantic"], "DLT Helpers for for pydantic.")
+
+
+def pydantic_to_table_schema_columns(model: Union[BaseModel, Type[BaseModel]]) -> TTableSchemaColumns:
```
Comment (rudolfix) [inline dlt/common/libs/pydantic.py:14]: this is good. nullability is set and code to infer type reused :)

could we add one optional parameter: `skip_complex_types` default to False. if set we do not add columns with complex types at all to generated schema. that would mimic the default dlt behaviour that creates child tables.

File: dlt/extract/utils.py [hunk 6]
```diff
@@ -12,3 +20,36 @@ def resolve_column_value(column_hint: TTableHintTemplate[TColumnNames], item: TD
...
+
+    raise ValueError(f"Unsupported columns type: {type(columns)}")
+
+
+def ensure_table_schema_columns_hint(columns: TTableHintTemplate[TAnySchemaColumns]) -> TTableHintTemplate[TTableSchemaColumns]:
+    """Convert column schema hint to a hint returning `TTableSchemaColumns`.
+    A callable hint is wrapped in another function which converts the original result.
+    """
+    if callable(columns) and not isinstance(columns, type):
+        def wrapper(item: TDataItem) -> TTableSchemaColumns:
```
Comment (rudolfix) [inline dlt/extract/utils.py:51]: forgot about that :) could you test it?

**Review by rudolfix — CHANGES_REQUESTED** [review]
thx for unifying the `columns` data type (we had both dicts and lists in different methods) and all the tests.

besides a few small comments in the review: could you add a short explanation to `resources.md` below `### Define schema` ie. `### Define schema with pydantic` where you can give an example and describe behavior (ie. lists, dics -> complex type/no child tables. optional - not null, unions - first element)

thanks!

**Comment by steinitzu** [comment]
Hey @rudolfix , added some docs, `skip_complex_types` and few more tests

**Review by rudolfix — APPROVED** [review]
excellent

**Pull request merged**